pkg/generator/secret_template_reconciler.go

```diff
@@ -0,0 +1,272 @@
+}
+
+// Reconcile is the entrypoint for incoming requests from k8s
+func (r *SecretTemplateReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithValues("request", request)
+
+	secretTemplate := sg2v1alpha1.SecretTemplate{}
+	if err := r.client.Get(ctx, types.NamespacedName{Namespace: request.Namespace, Name: request.Name}, &secretTemplate); err != nil {
+		if errors.IsNotFound(err) {
+			log.Info("Not found")
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	if secretTemplate.DeletionTimestamp != nil {
+		return reconcile.Result{}, nil
+	}
+
+	status := &reconciler.Status{
+		S:          secretTemplate.Status.GenericStatus,
+		UpdateFunc: func(st sgv1alpha1.GenericStatus) { secretTemplate.Status.GenericStatus = st },
+	}
+
+	status.SetReconciling(secretTemplate.ObjectMeta)
+	defer r.updateStatus(ctx, &secretTemplate)
+
+	res, err := r.reconcile(ctx, &secretTemplate)
+	//TODO is this overly defensive?
+	if err != nil {
+		secretTemplate.Status.Secret.Name = ""
+		if deleteErr := r.deleteChildSecret(ctx, &secretTemplate); deleteErr != nil {
+			return status.WithReconcileCompleted(res, deleteErr)
+		}
+	}
+
+	return status.WithReconcileCompleted(res, err)
+}
+
+func (r *SecretTemplateReconciler) reconcile(ctx context.Context, secretTemplate *sg2v1alpha1.SecretTemplate) (reconcile.Result, error) {
+	//Resolve input resources
+	inputResources, err := r.resolveInputResources(ctx, secretTemplate)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	//Template Secret Data
+	secretData := map[string][]byte{}
+	for key, expression := range secretTemplate.Spec.JSONPathTemplate.Data {
+		valueBuffer, err := JSONPath(expression).EvaluateWith(inputResources)
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("templating data: %w", err)
+		}
+
+		decoded, err := base64.StdEncoding.DecodeString(valueBuffer.String())
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("failed decoding base64 from a Secret: %w", err)
+		}
+
+		secretData[key] = decoded
+	}
+
+	//Template Secret StringData
+	secretStringData := map[string]string{}
+	for key, expression := range secretTemplate.Spec.JSONPathTemplate.StringData {
+		valueBuffer, err := JSONPath(expression).EvaluateWith(inputResources)
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("templating stringData: %w", err)
+		}
+
+		secretStringData[key] = valueBuffer.String()
+	}
+
+	//Create/Update Secret
+	secret := corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      secretTemplate.GetName(),
+			Namespace: secretTemplate.GetNamespace(),
+		},
+	}
+
+	if _, err = controllerutil.CreateOrUpdate(ctx, r.client, &secret, func() error {
+		secret.ObjectMeta.Labels = secretTemplate.Spec.JSONPathTemplate.Metadata.Labels
+		secret.ObjectMeta.Annotations = secretTemplate.Spec.JSONPathTemplate.Metadata.Annotations
```

Comment:
This is totally possible, do you have any specific usecases in mind? We're open to adding this.

In general, perhaps all values within `JSONPathTemplate` allow templating. Practically this would mean adding templating to `type`, `labels` and `annotations.

Reply:
> This is totally possible, do you have any specific usecases in mind?

if for whatever reason annotation need to be carried over from some input resource. but generally, yeah, seems a bit awkward if not all fields go through same "mutation"

Reply:
Should now have the ability to have templating in `type`, `labels` and `annotations`.